### DISCUSSION TITLE
GH API to update branch permissions

### DISCUSSION THREAD

**Discussion by ydorbala** [community/community#24758] [Apps, API and Webhooks]
<p>I am trying to update a branch permission by using gh apis. I authenticated my gh terminal using <code>gh auth login</code> and am able to execute certain gh commands as well.</p>
<p>However, when I am trying to update branch permissions, I am seeing Invalid request errors. I tried running the same using curl and this request is successful. How do I replicate the same using gh api?</p>
<pre><code class="lang-auto">curl --location --request PUT 'https://api.github.com/repos/&lt;orgName&gt;/&lt;repoName&gt;/main/protection' \
--header 'Authorization: Bearer &lt;token&gt; \
--header 'Content-Type: application/json' \
--data-raw '{"required_status_checks":{"strict":true,"contexts":["contexts"]},"enforce_admins":true,"required_pull_request_reviews": {},"restrictions":{"users":["users"],"teams":["teams"],"apps":["apps"]}}'
</code></pre>
<p>This is what I have come up with thus far (I tried putting in the same body as above but that was failing as well):</p>
<pre><code class="lang-auto">gh api -X PUT repos/&lt;orgName&gt;/&lt;repoName&gt;/branches/main/protection -f enforce_admins='true' -f required_pull_request_reviews='{}' -f restrictions='{}' -f required_status_checks='{}'` 
</code></pre>
<p>and the error I am seeing looks like this.</p>
<pre><code class="lang-auto">{
  "message": "Invalid request.\n\nNo subschema in \"anyOf\" matched.\nNo subschema in \"anyOf\" matched.\nNo subschema in \"anyOf\" matched.\nNo subschema in \"anyOf\" matched.\nFor 'allOf/0', \"{}\" is not an object.\nNot all subschemas of \"allOf\" matched.\nFor 'anyOf/1', \"{}\" is not a null.\nFor 'allOf/0', \"{}\" is not an object.\nNot all subschemas of \"allOf\" matched.\nFor 'anyOf/1', \"{}\" is not a null.\nFor 'allOf/0', \"true\" is not a boolean.\nNot all subschemas of \"allOf\" matched.\nFor 'anyOf/1', \"true\" is not a null.\nFor 'allOf/0', \"{}\" is not an object.\nNot all subschemas of \"allOf\" matched.\nFor 'anyOf/1', \"{}\" is not a null.",
  "documentation_url": "https://docs.github.com/rest/reference/repos#update-branch-protection"
}
gh: Invalid request.
</code></pre>
<p>Any help on how to send the right request through gh api would be really helpful. Thanks!</p>

**Comment by ydorbala** [comment, marked as answer]
<p>This post helped me answer my question. So, I had to pass in the array object as input as opposed to the field value (as that cannot take array values currently)</p>
<p>Something like this worked for me:</p>
<pre><code class="lang-auto">jq -n '{"required_status_checks":{"strict":true,"contexts":["contexts"]},"enforce_admins":true,"required_pull_request_reviews": {},"restrictions":{"users":["users"],"teams":["teams"],"apps":["apps"]}}' | gh api -X PUT repos/&lt;orgName&gt;/&lt;repoName&gt;/branches/main/protection --input -
</code></pre>

**Comment by joshjohanning** [comment]
Something like this might also work

```
gh api -X PUT /repos/ORG/REPO/branches/main/protection \
  --input - << EOF
{
  "required_status_checks": null,
  "enforce_admins": null,
  "required_pull_request_reviews": {
    "dismiss_stale_reviews": false,
    "require_code_owner_reviews": false,
    "require_last_push_approval": false,
    "required_approving_review_count": 1
  },
  "restrictions": null
}
```

**Comment by pgangloff-hbs** [comment]
Yep! @ydorbala 's solution above works for me too!

```
echo "Enabling main branch protection"
jq -n '{"required_status_checks":null,"enforce_admins":null,"required_pull_request_reviews":{"dismissal_restrictions":{},"required_approving_review_count":0},"dismiss_stale_reviews":null,"require_code_owner_reviews":null,"require_last_push_approval":null,"bypass_pull_request_allowances":null,"restrictions":null,"required_linear_history":null,"allow_force_pushes":null,"allow_deletions":null,"block_creations":null,"required_conversation_resolution":null,"lock_branch":null,"allow_fork_syncing":null}' \
| gh api -X PUT repos/MY_ORG/MY_PROJECT/branches/main/protection --input -
```

FYI: because I struggled with this, this is how we can disable branch protection programmatically (set it to null)

```
jq -n '{"required_status_checks":null,"enforce_admins":null,"required_pull_request_reviews":null,"dismiss_stale_reviews":null,"require_code_owner_reviews":null,"require_last_push_approval":null,"bypass_pull_request_allowances":null,"restrictions":null,"required_linear_history":null,"allow_force_pushes":null,"allow_deletions":null,"block_creations":null,"required_conversation_resolution":null,"lock_branch":null,"allow_fork_syncing":null}' \
| gh api -X PUT repos/MY_ORG/MY_PROJECT/branches/main/protection --input -
```
